### PR TITLE
[LOG-1815] Enhancement proposal: Add alerts and rules for operator-managed LokiStack

### PR DESCRIPTION
**Pull request by ronensc**
https://issues.redhat.com/browse/LOG-1815

### CODE REVIEW

File: enhancements/cluster-logging/loki-observability.md [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+### User Stories
+
+### Implementation Details
+
+The operational alerts are used to send alerts from Loki deployment indicating some level of unhealthy or misbehavior of
+the Loki Deployment. The alerts are based on metrics consumed from Prometheus into alert manager, evaluated and when 
+fired, pushed to notify human operators.
+
+Loki Alerts can be categorized as follows:
```
[the remark this answers is not in the document]
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md:51]: Done

File: enhancements/cluster-logging/loki-observability.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+* **Warning**: No data loss, but attention is required. This might be both, information before error or some temporal warning
+* **Info**: Information only alert
+
+#### Alerts
+
+Some alerts are based on metrics exposed by Loki and some on recording rules, which are metrics created based on other metrics.
+
+See also [Defining recording rules](https://prometheus.io/docs/prometheus/latest/configuration/recording_rules/)
+
+##### LokiRequestErrors
```
[the remark this answers is not in the document]
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md]: Done

File: enhancements/cluster-logging/loki-observability.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+### User Stories
+
+### Implementation Details
+
+The operational alerts are used to send alerts from Loki deployment indicating some level of unhealthy or misbehavior of
+the Loki Deployment. The alerts are based on metrics consumed from Prometheus into alert manager, evaluated and when 
+fired, pushed to notify human operators.
+
+Loki Alerts can be categorized as follows:
+* **Loki System**: One or more of the Loki components is misbehaving; this can be due to any of Loki or OCP system that is not working as expected
```
[the remark this answers is not in the document]
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md]: I installed `loki-operator` and checked the labels of `loki_panic_total` metric. It seems that we can distinguish between loki components using the `job` label
image

File: enhancements/cluster-logging/loki-observability.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+Note: `loki_panic_total` metric is implemented but not documented.
+* https://github.com/grafana/loki/blob/a046d7926122cef1b619d89bfcf15403e14e0de3/pkg/util/server/recovery.go#L19-L23
+* https://grafana.com/docs/loki/latest/operations/observability/
+
+References:
+* https://monitoring.mixins.dev/loki/
+* https://github.com/rhobs/configuration/blob/main/docs/sop/observatorium.md#lokirequestpanics
+
+##### LokiRequestLatency
```
[the remark this answers is not in the document]
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md:131]: All tenants I guess. I couldn't find a label that distinguishes between labels

File: enhancements/cluster-logging/loki-observability.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+> 100
+```
+
+Note: `loki_discarded_samples_total` metric is implemented but not documented
+* https://github.com/grafana/loki/blob/25163470b0b861d5c539b66a4c5613e46666e0c8/pkg/validation/validate.go#L98-L105
+* https://grafana.com/docs/loki/latest/operations/observability/
+
+References:
+* https://github.com/rhobs/configuration/blob/main/docs/sop/observatorium.md#lokitenantratelimitwarning
+
```
[the remark this answers is not in the document]
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md:175]: I added a summary table at the beginning and added a few more alerts

File: enhancements/cluster-logging/loki-observability.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+* https://github.com/grafana/loki/blob/a046d7926122cef1b619d89bfcf15403e14e0de3/pkg/util/server/recovery.go#L19-L23
+* https://grafana.com/docs/loki/latest/operations/observability/
+
+References:
+* https://monitoring.mixins.dev/loki/
+* https://github.com/rhobs/configuration/blob/main/docs/sop/observatorium.md#lokirequestpanics
+
+##### LokiRequestLatency
+
+The 99th percentile is experiencing high latency.
```
[the remark this answers is not in the document]
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md]: done

File: enhancements/cluster-logging/loki-observability.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+Category: Loki System
+
+Severity: critical
+
+query:
+```
+namespace_job_route:loki_request_duration_seconds:99quantile{route!~"(?i).*tail.*"} > 1
+```
+
+underline recording rule:
```
[the remark this answers is not in the document]
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md]: done

File: enhancements/cluster-logging/loki-observability.md [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+Alerts summary table:
+
+| Alert | Category | Severity | Summary |
+| ----- | -------- | -------- | ----------- |
+| [LokiRequestErrors](#LokiRequestErrors) | Loki System | Critical | At least 10% of requests are responded by 5xx server errors |
+| [LokiRequestPanics](#LokiRequestPanics) | Loki System | Critical | A panic was triggered |
+| [LokiRequestLatency](#LokiRequestLatency) | Loki System | Critical | The 99th percentile is experiencing high latency (higher than 1 second) |
+| [LokiTenantRateLimitWarning](#LokiTenantRateLimitWarning) | Tenant Workload | Warning | A tenant is experiencing rate limiting |
+| [LokiCPUHighWarning](#LokiCPUHighWarning) | Loki System | Warning | Loki's CPU usage is high (above 75% for 5 minutes) |
+| [LokiCPUHighError](#LokiCPUHighError) | Loki System | Critical | Loki's CPU usage is high (above 95% for 5 minutes) |
```
Comment (alanconway) [inline enhancements/cluster-logging/loki-observability.md:85]: These can both be called LokiCPUHigh - you can issue the same alert with a different severity based on different rules. You only do that for different intensities of "the same condition", which is the case here.  [An example of this pattern](https://github.com/openshift/cluster-monitoring-operator/blob/fd2cbc333d5a89c816db3f5dead99dec6d1e6a7f/assets/node-exporter/prometheus-rule.yaml#L17-L52)
```suggestion
| [LokiCPUHigh](#LokiCPUHigh) | Loki System | Warning | Loki's CPU usage is high (above 75% for 5 minutes) |
| [LokiCPUHigh](#LokiCPUHigh) | Loki System | Critical | Loki's CPU usage is high (above 95% for 5 minutes) |
```

File: enhancements/cluster-logging/loki-observability.md [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+| [LokiMemoryHighWarning](#LokiMemoryHighWarning) | Loki System | Warning | Loki's memory usage is high (above 75% of the configured limit for 5 minutes) |
+| [LokiMemoryHighError](#LokiMemoryHighError) | Loki System | Critical | Loki's memory usage is high (above 95% of the configured limit for 5 minutes) |
```
Comment (alanconway) [inline enhancements/cluster-logging/loki-observability.md:87]: Same comment, both can be LokiMemoryHigh. In general, don't suffix alert names with a severity like "FooWarning", the severity is delivered as part of the alert, it shouldn't be in the name.

**Review by alanconway — CHANGES_REQUESTED** [review]
Suggested some minor name changes, otherwise LGTM.

File: enhancements/cluster-logging/loki-observability.md (hunk 9 above)
Comment (ronensc) [reply enhancements/cluster-logging/loki-observability.md:87]: Great feedback. thanks

**Comment by ronensc** [comment]
@alanconway @periklis @eranra 
Since Loki is just a logging system, is there an alert that is justified a `critical` severity? (i.e. warrants waking someone in the middle of the night)

**Review by alanconway — APPROVED** [review]
Need to squash commits before merge, otherwise LGTM.

**Review by periklis — APPROVED** [review]
/lgtm

**Pull request merged**